controller/api/destination/watcher/server_watcher.go

```diff
@@ -0,0 +1,148 @@
+package watcher
+
+import (
+	"sync"
+
+	"github.com/linkerd/linkerd2/controller/gen/apis/server/v1beta1"
+	"github.com/linkerd/linkerd2/controller/k8s"
+	logging "github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/tools/cache"
+)
+
+// ServerWatcher watches all the servers in the cluster. When there is an
+// update, it will update any listeners for pods that are selected by the
+// update's Server.
+type ServerWatcher struct {
+	subscriptions map[podPort][]ServerUpdateListener
+	k8sAPI        *k8s.API
+	log           *logging.Entry
+	sync.RWMutex
+}
+
+type podPort struct {
+	pod  *corev1.Pod
+	port Port
+}
+
+// ServerUpdateListener is the interface that subscribers must implement.
+type ServerUpdateListener interface {
+	Update(bool)
```

Comment:
I think this could use documentation of the interface.  Namely that this only informs the listener of the opaqueness of the pod+port instead of on any changes to a matching server.